lib/autoupdate.ps1

```diff
@@ -431,3 +424,104 @@ function autoupdate([String] $app, $dir, $json, [String] $version, [Hashtable] $
         Write-Host -f DarkGray "No updates for $app"
     }
 }
+
+## Helper Functions
+
+function PropertyHelper {
+    <#
+    .SYNOPSIS
+        Helper of updating property
+    .DESCRIPTION
+        Update manifest property (String, Array or PSCustomObject).
+    .PARAMETER Property
+        Property to be updated
+    .PARAMETER Value
+        New property values
+        Update line by line
+    .OUTPUTS
+        System.Object[]
+            The first element is new property, the second element is change flag
+    #>
+    param (
+        [Object]$Property,
+        [Object]$Value
+    )
+    $hasChanged = $false
+    switch ($Property.GetType().Name) {
+        'String' {
+            $Value = $Value -as [String]
+            if (($null -ne $Value) -and ($value -ne $Property)) {
+                $Property = $Value
+                $hasChanged = $true
+            }
+        }
+        'Object[]' {
+            $Value = @($Value)
+            for ($i = 0; $i -lt [Math]::Min($Property.Length, $Value.Length); $i++) {
+                $Property[$i], $hasItemChanged = PropertyHelper -Property $Property[$i] -Value $Value[$i]
+                $hasChanged = $hasChanged -or $hasItemChanged
+            }
+        }
+        'PSCustomObject' {
+            if ($Value -is [PSObject]) {
+                foreach ($name in $Property.PSObject.Properties.Name) {
+                    if ($Value.$name) {
+                        $Property.$name, $hasItemChanged = PropertyHelper -Property $Property.$name -Value $Value.$name
+                        $hasChanged = $hasChanged -or $hasItemChanged
+                    }
+                }
+            }
+        }
+    }
+    return $Property, $hasChanged
+}
+
+function HashHelper {
+    <#
+    .SYNOPSIS
+        Helper of getting file hash(es)
+    .DESCRIPTION
+        Get file hash(es) by hash extraction template(s).
+        If hash extraction templates are less then URLs, the last template will be reused for the rest URLs.
+    .PARAMETER AppName
+        Software name
+    .PARAMETER Version
+        Given software version
+    .PARAMETER HashExtraction
+        Hash extraction template(s)
+    .PARAMETER URL
+        New download URL(s), used to calculate hash locally
```

Comment:
This is not accurate. It will be used for calculating hash locally or extracting hash.

> used to calculate hash locally

This implies it will be downloaded and calculated all the time, without possible hash extraction when defined and working.

Reply:
Yes please. Sounds good.